src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxValueProvider_ForAttributeWithSimpleName.cs

```diff
@@ -83,19 +87,23 @@ public partial struct SyntaxValueProvider
             .Select((tuple, _) => GlobalAliases.Concat(tuple.Left, tuple.Right))
             .WithTrackingName("allUpIncludingCompilationGlobalAliases_ForAttribute");
 
-        // Combine the two providers so that we reanalyze every file if the global aliases change, or we reanalyze a
-        // particular file when it's compilation unit changes.
-        var syntaxTreeAndGlobalAliasesProvider = syntaxTreesProvider
-            .Combine(allUpGlobalAliasesProvider)
-            .WithTrackingName("compilationUnitAndGlobalAliases_ForAttribute");
+        // Filter down to the trees that contain attributes in them.  in general, this will normally be smaller than the
+        // number of total trees, allowing us to hold onto a much smaller number of nodes.
+        var treesContainingAttribute = syntaxTreesProvider.Where(
+            (t, c) =>
+            {
+                // Walk the green node tree first to avoid allocating the entire red tree for files that have no attributes.
+                var root = t.GetRoot(c);
+                return ContainsAttributeList(root.Green, syntaxHelper.AttributeListKind);
+            });
```

Comment:
this part is substantial.  Previously, we would always join all trees with the global-attributes-info.  this meant on a chnage to any tree, we had to rebuild this table.  now, we only join the trees that have attributes with teh global-attributes-info. this means, if you edit a tree without an attribute in it, we produce the same list of trees+global-attributes-info, and we can reuse that table entirely (which also prevents any further downstream work).

We could also extend this in teh future to run the predicate here.  that way we would both only check trees with attributes *and* which match the user predicate, protentially reducing this down further.